ops/pebble.py

```diff
@@ -0,0 +1,537 @@
+
+    def __init__(self, socket_path):
+        super(_UnixSocketHandler, self).__init__()
+        self.socket_path = socket_path
+
+    def http_open(self, req):
+        return self.do_open(_UnixSocketConnection, req, socket_path=self.socket_path)
+
+
+def _fromisoformat(s):
+    """Convert datetime string in ISO format to a datetime object.
+
+    Equivalent to datetime.fromisoformat (for our purposes), but that was only
+    introduced in Python 3.7, and we need to support Python 3.5.
+    """
+    return datetime.datetime.strptime(s, '%Y-%m-%dT%H:%M:%S.%f%z')
+
+
+_TIMESTAMP_RE = re.compile(r'(.*)\.(\d+)(.*)')
+
+
+def _parse_timestamp(s):
+    """Parse timestamp from Go-encoded JSON (which uses 9 decimal places for seconds)."""
+    match = _TIMESTAMP_RE.match(s)
+    if not match:
+        return _fromisoformat(s)
+    head, subsecond, rest = match.groups()
+    subsecond = subsecond[:6]  # fromisoformat supports at most 6 decimal places
+    return _fromisoformat(head + '.' + subsecond + rest)
+
+
+class ServiceError(Exception):
+    """Raised by API.wait_change when a service change is ready but has an error."""
+
+    def __init__(self, err, change):
+        self.err = err
+        self.change = change
+
+    def __str__(self):
+        return self.err
+
+
+class WarningState(enum.Enum):
+    """Enum of states for API.get_warnings() select parameter."""
+
+    ALL = 'all'
+    PENDING = 'pending'
+
+
+class ChangeState(enum.Enum):
+    """Enum of states for API.get_changes() select parameter."""
+
+    ALL = 'all'
+    IN_PROGRESS = 'in-progress'
+    READY = 'ready'
+
+
+class SystemInfo:
+    """System information object."""
+
+    def __init__(self, version: str):
+        self.version = version
+
+    @classmethod
+    def from_dict(cls, d: Dict) -> 'SystemInfo':
+        """Create new object from dict parsed from JSON."""
+        return cls(version=d['version'])
+
+    def __repr__(self):
+        return 'SystemInfo(version={self.version!r})'.format(self=self)
+
+    __str__ = __repr__
+
+
+class Warning:
+    """Warning object."""
+
+    def __init__(
+        self,
+        message: str,
+        first_added: datetime.datetime,
+        last_added: datetime.datetime,
+        last_shown: Optional[datetime.datetime],
+        expire_after: str,
+        repeat_after: str,
+    ):
+        self.message = message
+        self.first_added = first_added
+        self.last_added = last_added
+        self.last_shown = last_shown
+        self.expire_after = expire_after
+        self.repeat_after = repeat_after
+
+    @classmethod
+    def from_dict(cls, d: Dict) -> 'Warning':
+        """Create new object from dict parsed from JSON."""
+        return cls(
+            message=d['message'],
+            first_added=_parse_timestamp(d['first-added']),
+            last_added=_parse_timestamp(d['last-added']),
+            last_shown=_parse_timestamp(d['last-shown']) if d.get('last-shown') else None,
```

Comment:
`.get` does exactly that, just do:

    last_shown=_parse_timestamp(d.get('last-shown')),

Reply:
No, that would fail because we'd be passing `None` to `_parse_timestamp()` if the field wasn't present (or was None), which would fail as that requires a string. Note the placement of the parentheses -- we're only calling `_parse_timestamp` if the `last-shown` value is set and not None.

Reply:
You're right, I was misreading the closing parenthesis.